trtllm-truss/README.md

```diff
@@ -0,0 +1,88 @@
+
+### Prerequisites 
+
+To use this Truss, your engine must be built with in-flight batching support. Refer to your architecture-specific `build.py` re: how to build with in-flight-batching support. 
+
+### Config
+
+This Truss is primarily config driven. This means that most settings you'll need to edit are located in the `config.yaml`. These settings are all located underneath the `model_metadata` key. 
+
+- `tensor_parallelism` (int): If you built your model with tensor parallelism support, you'll need to set this value with the same value used during the build engine step. This value should be the same as the number of GPUs in the `resources` section. 
+
+*Pipeline parallelism is not supported in this version but will be added later. As noted from Nvidia, pipeline parallelism reduces the need for high-bandwidth communication but may incur load-balancing issues and may be less efficient in terms of GPU utilization.*
+
+- `engine_repository` (str): We expect engines to be uploaded to Huggingface with a flat directory structure (i.e the engine and associated files are not underneath a folder structure). This value is the full `{org_name}/{repo_name}` string. Engines can be private or public.
+
+- `tokenizer_repository` (str): Engines do not come bundled with their own tokenizer. This is the Huggingface repository where we can find a tokenizer. Tokenizers can be private or public.
+
+If the engine and repository tokenizers are private, you'll need to update the `secrets` section of the `config.yaml` as follows:
+
+```
+secrets:
+ hf_access_token: "my_hf_api_key"
+```
+
+### Performance 
+
+TRT-LLM engines are designed to be highly performant. Once your Truss has been deployed, you may find that you're not fully utilizing the GPU. The following are levers to improve performance but require trial-and-error to identify appropriates. All of these values live inside the `config.pbtxt` for a given ensemble model. 
+
+#### Preprocessing / Postprocessing
+
+```
+instance_group [
+    {
+        count: 1
+        kind: KIND_CPU
+    }
+]
+```
+By default, we load 1 instance of the pre/post models. If you find that the tokenizer is a bottleneck, increasing the `count` variable here will load more replicas of these models and Triton will automatically load balance across model instances. 
+
+### Tensorrt LLM
+```
+parameters: {
+  key: "max_tokens_in_paged_kv_cache"
+  value: {
+    string_value: "10000"
+  }
+}
+```
+By default, we set the `max_tokens_in_paged_kv_cache` to 10000. For a 7B model on 1 A100 with a batch size of 8, we have over 60GB of GPU memory left over. We can increase this value to 100k comfortably and allow for more tokens in the KV cache. Your mileage will vary based on the size of your model and the hardware you're running on. 
+
+```
+parameters: {
+  key: "kv_cache_free_gpu_mem_fraction"
+  value: {
+    string_value: "0.1"
+  }
+}
+```
+TODO(Abu): __fill__
```

Comment:
for your [reference](https://github.com/triton-inference-server/tensorrtllm_backend/blob/47b609b670d6bb33a5ff113d98ad8a44d961c5c6/inflight_batcher_llm/src/libtensorrtllm.cc#L1113): looks like using both `max_tokens_in_paged_kv_cache` and `kv_cache_free_gpu_mem_fraction` are redundant. I didn't find docs on `kv_cache_free_gpu_mem_fraction` but sounds like they preallocate 85% of the free gpu mem for the kv cache by default, if `max_tokens_in_paged_kv_cache` is not specified